src/backend/distributed/operations/shard_rebalancer.c

```diff
@@ -2609,7 +2611,17 @@ ActivePlacementsHash(List *shardPlacementList)
 	foreach(shardPlacementCell, shardPlacementList)
 	{
 		ShardPlacement *shardPlacement = (ShardPlacement *) lfirst(shardPlacementCell);
-		if (shardPlacement->shardState == SHARD_STATE_ACTIVE)
+		WorkerNode *workerNode = FindWorkerNode(shardPlacement->nodeName,
```

Comment:
how can we have NULL worker node? Does that mean the worker removed from the cluster?

Reply:
When you run smth like this:
```sql
CREATE OR REPLACE FUNCTION shard_placement_rebalance_array(
    worker_node_list json[],
    shard_placement_list json[],
    threshold float4 DEFAULT 0,
    max_shard_moves int DEFAULT 1000000,
    drain_only bool DEFAULT false,
    improvement_threshold float4 DEFAULT 0.5
)
RETURNS json[]
AS 'citus'
LANGUAGE C STRICT VOLATILE;

-- Check that even with threshold=0.0 shard_placement_rebalance_array returns
-- something when there's no completely balanced solution.


SELECT unnest(shard_placement_rebalance_array(
    ARRAY['{"node_name": "hostname1"}',
          '{"node_name": "hostname2"}']::json[],
    ARRAY['{"shardid":1, "nodename":"hostname1"}',
          '{"shardid":2, "nodename":"hostname1"}',
          '{"shardid":3, "nodename":"hostname1"}']::json[]
));
```

`FindWorkerNode` attempts to find the workers of the shards and does not find `hostname1` or `hostname2` in the actual worker nodes.

Reply:
When I saw `workerNode == NULL`, I expected it'd be qualified as !isActive. Or, it can actually never be NULL, because `ResolveGroupShardPlacement` function which is always used to load shardPlacements ensures that via `LookupNodeForGroup()`. If a placement is on a node that doesn't exist, we'd get:
```C
	if (!foundAnyNodes)
	{
		ereport(ERROR, (errmsg("there is a shard placement in node group %d but "
							   "there are no nodes in that group", groupId)));
	}
```


So, we seem to be doing this change for the sake of a fake test with a fake metadata.

I think we should probably insert these dummy host-names to pg_dist_node in the tests, like:
```SQL
BEGIN;
INSERT INTO pg_dist_node ...(hostname1) ;;

 SELECT unnest(shard_placement_rebalance_array( ARRAY['{"node_name": "hostname1"}', '{"node_name": "hostname2"}']::json[], ARRAY['{"shardid":1, "nodename":"hostname1"}', '{"shardid":2, "nodename":"hostname1"}', '{"shardid":3, "nodename":"hostname1"}']::json[] ));

ROLLBACK;
```

Such that they become more realistic. And, we can assert `workerNode != NULL` as described above (e.g., we cannot have placements that is on a non-existing node).